### ISSUE TITLE
SCSS file paths with custom base-href are wrong

### ISSUE THREAD

**Issue by btxtiger** [angular/angular-cli#23104]
# 🐞 Bug report

### Command (mark with an `x`)

- [x] build

### Is this a regression?
No, didn't work in previous versions as mentioned but not properly solved here #4782 or here #18043

## 🔬 Description & Minimal Reproduction
1) `ng new test-app`
2) Place a picture in `src/assets`

**_We have to look at 2 different build environments:_**

#### 1) Default dev env with `ng serve`
The default base-href is `<base href="/"></base>`, which means all URIs don't have a prefix and have the default root.

| Syntax      | Status |
| ----------- | ----------- |
|`<img src="/assets/img.png" />` | works, but ignores the base-href |
|`<img src="assets/img.png" />` | works, is relative path from the root |
|`background-imge: url('/assets/img.png')` | works, but ignores the base-href |
|`background-imge: url('assets/img.png')` | works not |
|`background-imge: url('~src/assets/img.png')` | works, why?? |

####  2) Serving a production build using a webserver like nginx or apache
Now we set the base-href to `<base href="/app/"></base>` in index.html and set in **angular.json** `"baseHref": "/app/",` in `architect/configurations/production` and run `ng build`.

| Syntax      | Status |
| ----------- | ----------- |
| `<img src="/assets/img.png" />`      | works not, because we have /app/ prefix after root       |
| `<img src="assets/img.png" />`   | works, path is updated to /app/assets/img.png by compiler        |
| `<img src="./assets/img.png" />`   | same as above        |
| `background-imge: url('/assets/img.png')`   |  works not, because we have /app/ prefix after root        |
| `background-imge: url('assets/img.png')`   | works not, build fails        |
| `background-imge: url('./assets/img.png')`   | works not, build fails        |
| `background-imge: url('~src/assets/img.png')`   | ~~works not~~   **update: works**     |
| `background-imge: url('~assets/img.png')`   | works not, build fails        |
| `background-imge: url('/app/assets/img.png')`   | works, but uses root routing and breaks dev env  |

**TLDR:**
There is no way to access asset files in SCSS with a changed base-href, without hardcoding the base-href in SCSS like `/app/assets/img.png` but breaking the local dev env.

## 🔥 Exception or Error

```sh
./src/app/elements/phone-frame/phone-frame.component.scss?ngResource - Error: Module Error (from ./node_modules/postcss-loader/dist/cjs.js):
/Users/me/repositories/app-frontend/src/app/elements/phone-frame/phone-frame.component.scss:16:28: Can't resolve './assets/images/phone-frame-dark.png' in '/Users/me/repositories/app-frontend/src/app/elements/phone-frame'
```

## 🌍 Your Environment

```sh
Angular CLI: 13.3.3
Node: 16.13.1
Package Manager: npm 8.1.2
OS: darwin x64

Angular: 13.3.4
... animations, cdk, common, compiler, compiler-cli, core, forms
... google-maps, material, platform-browser
... platform-browser-dynamic, router, service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1303.3
@angular-devkit/build-angular   13.3.3
@angular-devkit/core            13.3.3
@angular-devkit/schematics      13.3.3
@angular/cli                    13.3.3
@schematics/angular             13.3.3
rxjs                            6.6.7
typescript                      4.4.4
```

## Preferred solution
The CSS compiler should update the asset paths in CSS/SCSS/SASS files as like the `src` paths in HTML are updated.

**Comment by btxtiger** [comment]
After finding some server misconfiguration (for serving static assets), this syntax turned out to be definitely valid:
✅ `background-imge: url('~src/assets/img.png')`

However this feels very inconsistent and must be better documented aswell.

**Comment by alan-agius4** [comment]
Unlike assets in templates, relative resources in CSS, such as images and fonts, are automatically written and fingerprinted at the root of the output (dist) folder. This mean that these don't need to be added to the `assets` option in `angular.json`. 

In addition, unlike templates assets, Relative resources in CSS are relative to the CSS source file.

This is briefly mentioned in https://angular.io/cli/build

**Comment by alan-agius4** [comment]
Closing as per above.

**Comment by btxtiger** [comment]
@alan-agius4 
If the CSS assets were copied to root, what they are in deed, they should be accessible by setting the URL like this:
![image](https://user-images.githubusercontent.com/15804690/169553070-70412647-31cd-4f3d-86ed-da47e4e0cb32.png)

But this does not work with the ng compiler:
![image](https://user-images.githubusercontent.com/15804690/169553565-cbf922fa-c62b-4f17-b937-3af2848fd38d.png)

Afaik, this is the only working way, that works both, with `baseHref == '/'` and `baseHref == '/app/'`
![image](https://user-images.githubusercontent.com/15804690/169553762-d5561600-b8c2-4244-b9fd-9e9884f93c5e.png)

**Comment by alan-agius4** [comment]
For the first case to work, you need to place the resource relative to the component stylesheet. In this case, `photo-frame.png` needs to be in the same directory of the CSS file.

Relative stylesheet resources, don’t need to be copied using the `asset` option in the `angular.json` as these are processed and copied by the CSS build pipeline.

**Comment by btxtiger** [comment]
@alan-agius4 Okay, so the last case using `~src/assets/` is the way to go, if I want to keep the image files sorted in the image dir. Or am I getting something wrong?

**Comment by alan-agius4** [comment]
You can also resolve the path relative paths example `../../resources/…`.

Make sure that you are not coping the assets twice. In general I would recommend renaming the folder to something other than `assets`, since unless removed from the `angular.json` , files in the assets directory will also be copies
